packages/protocol/contracts/tokenvault/ERC20Vault.sol

```diff
@@ -217,58 +171,42 @@ contract ERC20Vault is
         });
     }
 
-    /// @notice Releases deposited ERC20 tokens back to the user on the source
-    /// ERC20Vault with a proof that the message processing on the destination
-    /// Bridge has failed.
-    /// @param message The message that corresponds to the ERC20 deposit on the
-    /// source chain.
-    function onMessageRecalled(IBridge.Message calldata message)
+    function onMessageRecalled(
+        IBridge.Message calldata message,
+        bytes32 msgHash
+    )
         external
         payable
         override
         nonReentrant
-        onlyFromNamed("bridge")
+        whenNotPaused
     {
-        IBridge bridge = IBridge(resolve("bridge", false));
-        bytes32 msgHash = bridge.hashMessage(message);
+        checkRecallMessageContext();
 
-        (, address token,, uint256 amount) = abi.decode(
-            message.data[4:], (CanonicalERC20, address, address, uint256)
-        );
+        (, address token,, uint256 amount) =
+            abi.decode(message.data[4:], (CanonicalERC20, address, address, uint256));
 
         if (token == address(0)) revert VAULT_INVALID_TOKEN();
 
         if (amount > 0) {
-            if (isBridgedToken[token] || token == resolve("taiko_token", true))
-            {
+            if (bridgedToCanonical[token].addr != address(0)) {
```

Comment:
@Brechtpd @dantaik @davidtaikocha @cyberhorsey 
I think (but please double check) i found a bug here - (during developing the native token support).


So onMessageRecall() releases the tokens on the "source" chain - if something happens (and result in `FAILED` status) on the "destination" chain.

If this is the case, why we `.burn()` from the ERC20Vault ? We should mint them, since the sendToken() already burnt the amount - if this is a bridged token.

I think we should `.mint()` not `.burn()`. 🤔  ?

Reply:
(line 193)

Reply:
https://github.com/taikoxyz/taiko-mono/pull/15272